apps/labs/posts/numpy-quaddtype-blog.md

```diff
@@ -0,0 +1,419 @@
+
+
+This flexibility allows `numpy_quaddtype` to provide the best available precision across different platforms while maintaining a consistent interface.
+
+By addressing these goals, `numpy_quaddtype` not only solves the immediate issues surrounding `np.longdouble` but also paves the way for more robust and precise numerical computing in Python. In the following sections, we'll delve deeper into how these goals are achieved through the technical implementation of `numpy_quaddtype`.
+
+
+### The Inner Workings of `numpy_quaddtype`
+
+`numpy_quaddtype` is built on a flexible architecture that marries high precision with cross-platform compatibility. At its core lies the `QuadPrecisionObject`, a chameleon-like structure that can switch between two forms:
+
+```c
+typedef union {
+    Sleef_quad sleef_value;
+    long double longdouble_value;
+} quad_value;
+
+typedef struct {
+    PyObject_HEAD
+    quad_value value;
+    QuadBackendType backend;
+} QuadPrecisionObject;
+```
+
+This dual-nature allows us to leverage the full 128-bit precision of SLEEF when available, while maintaining backwards compatibility with `np.longdouble`.
+
+To integrate seamlessly with NumPy, we've created the `QuadPrecDTypeObject`:
+
+```c
+typedef struct {
+    PyArray_Descr base;
+    QuadBackendType backend;
+} QuadPrecDTypeObject;
+```
+
+This structure acts as a bridge, allowing our high-precision numbers to work harmoniously within NumPy arrays and operations.
+
+The backend selection is handled by a simple enum:
+
+```c
+typedef enum {
+    BACKEND_INVALID = -1,
+    BACKEND_SLEEF,
+    BACKEND_LONGDOUBLE
+} QuadBackendType;
+```
+
+This flexibility enables `numpy_quaddtype` to provide optimal precision across different platforms while maintaining a consistent interface. Users can select their preferred `backend` at runtime:
+
+```python
+import numpy as np
+import numpy_quaddtype as npq
+
+# Using SLEEF backend (default)
+x = npq.QuadPrecision(3.14159)
+x =npq.QuadPrecision(3.14159, backend='sleef')
+
+# Using longdouble backend
+y = npq.(2.71828, backend='longdouble')
+
+# Creating a NumPy array with QuadPrecision dtype
+z = np.array([x, y], dtype=npq.QuadPrecDType()) # SLEEF
+z = np.array([x, y], dtype=npq.QuadPrecDType("longdouble")) # SLEEF
+```
+
+Under the hood, `numpy_quaddtype` manages memory efficiently for both **aligned** and **unaligned**. This is crucial for performance, especially when dealing with large arrays or complex computations. We've implemented specialized strided loop functions for various operations
+
+<figure style={{ textAlign: 'center' }}>
+    <img 
+      src="/posts/numpy-quaddtype-blog/aligned_unaligned_memory.png"
+      alt="Diagram illustrating an example of aligned versus unaligned memory"
+      style={{ display: 'inline-block', maxWidth: '100%', height: 'auto' }}
+    />
+    <figcaption>Figure 1: Illustrating an example of aligned and unaligned memory</figcaption>
+</figure>
```

Comment:
So this is an interesting point, but I think the way you have it presented here is confusing. 

I suggested this earlier on slack and after reading this in context, I think if you want to keep this discussion, you should first explain how structs can have aligned or un-aligned layouts using the existing image. It's worth having a second image showing how numpy arrays allow both aligned and unaligned access to data (e.g. by creating a structured dtype array that includes a quad dtype in the struct). Then, you would need to explain that it's possible to create arrays with structured dtypes where each element of the array is a struct, so NumPy needs to worry about this. You could have done the easy and slow thing, where you only implement the more general un-aligned case, but then you can show how that requires two memcpy operations that are unnecessary for the far more common aligned case.

Reply:
Another options is just to delete this discussion of aligned vs unaligned access, since it's pretty in the weeds IMO.

Reply:
Yeah make sense, I think, we just leave the point that dtype can support both aligned and unaligned memory access and remove the explaining of `what aligned and unaligned means`  (diagram basically)